### ISSUE TITLE
Simple example to parse metadata about functions for a binary

### ISSUE THREAD

**Issue by vsoch** [angr/angr#3125]
Hi angr team! I'm looking to do some simple binary analysis to get a listing of functions, and for each:

1. the name of the function
2. parameters and types, likely nested for structs, and more complex types
3. the register that parameter is intended for, per [system V abi document](https://refspecs.linuxbase.org/elf/x86_64-abi-0.99.pdf) 
4. parsing of call sites
5. *bonus* some kind of exception parsing

Is this something you can do and point me to a good example? I'm currently using https://github.com/dyninst/dyninst for some of these things (except for 3. that is added / attempted separately) but I find it kind of bug laden and am hoping I could use Python to write the same. Thank you!

**Comment by rhelmot** [comment]
Hello! Here are some resources for you:

1. Build a CFG (project.analyses.CFGFast()). This will recover functions and associate names from the binary's debug symbols with them.
2. Run the calling convention analysis (project.analyses.CallingConventionAnalysis()), then variable recovery (project.analyses.VariableRecoveryFast()), then type inference (project.analyses.Typehoon()).
3. The calling convention you get back from the first part of 2 will be a SimCC. This (when provided with a function prototype, which CCA should also give you) gives you very detailed information about how a function's parameters should be laid out in registers and memory.
4. Unclear what you're asking for. Various analyses listed above take callsites into consideration during their analysis, and you can retrieve all callsites of a function by asking for its predecessors in the interprocedural CFG, but I don't know what it means to "parse" a callsite.
5. CLE will automatically parse any debug information related to exception handling (for ELF tiles only) into project.loader.main_object.exception_handlings. CFGFast will by default use these to add special "exception" edges to the graph, though I'm not super clear on what information this gives you.

Hope this helps!

**Comment by rhelmot** [comment]
Oh yeah, I forgot to say - you should probably read through the source of the [decompiler analysis](https://github.com/angr/angr/blob/master/angr/analyses/decompiler/decompiler.py) as it is the canonical example for how to use most of these analyses.

**Comment by vsoch** [comment]
Wicked! Thank you! I'll try this out asap (today). Is it okay if I leave it open for today and possibly into the weekend in case I have any questions? I promise I will close as soon as I finish!

**Comment by rhelmot** [comment]
![image](https://user-images.githubusercontent.com/2498805/152577070-fabcec13-2608-4935-9bf8-72295d1762ab.png)

I don't think we care much about making sure issues get closed :) Leave it open until you're satisfied.

**Comment by vsoch** [comment]
holy cow this is awesome! So I think I found the registers - I'm looking at the result of the convention analysis and I see

```python
convention.cc.arch.argument_registers
{24, 32, 64, 72, 80, 88, 96}
```
But this seems like it's a data structure for ALL registers across the arch? The function I'm inspecting has 6 arguments and this shows 7, so I first was thinking maybe this is arguments plus the return value? is it correct that these are the argument registers for this specific function? Naively I also see register names:

```python
'register_names': {16: 'rax',
  24: 'rcx',
  32: 'rdx',
  40: 'rbx',
  48: 'rsp',
  56: 'rbp',
  64: 'rsi',
...
```
So if I were to look them up:

```python
for regnum in convention.cc.arch.argument_registers:
     ...:     print(convention.cc.arch.register_names[regnum])
rdx
rsi
r10
rdi
r8
rcx
r9
```
So actually I think it might be the case that one variable is really big and possibly taking up two registers? I also ran the variable analysis and types, and I'm trying to figure out how to map the output to what I know about the function (e.g., I know the first variable is named a and has size 8, is a long int, and should be in `%rdi`.

```
{
          "name":"a",
          "type":"long int",
          "class":"Integer",
          "location":"%rdi",
          "direction":"import",
          "size":"8"
        },
```
Thanks for your help! I (think?) everything is here that I'm looking for, just figuring out the conventions of things.

**Comment by vsoch** [comment]
ohh hang on I think I was looking at the wrong function - one second! I see something! <3

**Comment by vsoch** [comment]
I think I must remember incorrectly and the function does have 7 arguments, I now see:

```python
# Note there are 7 plus an integer return
convention.prototype
(long long, long long, long long, long long, long long, long long, long long) -> int
```
In which case this makes sense!
```python
for regnum in convention.cc.arch.argument_registers:
    print(convention.cc.arch.register_names[regnum])
...
rdx
rsi
r10
rdi
r8
rcx
r9
```
But it does appear the order is off, likely because the argument registers are in a set.

```
type(convention.cc.arch.argument_registers)
set
```

I think this might be what I need?

```python
 for regnum in convention.cc.arch.argument_registers:
    print(convention.cc.arch.argument_register_positions[regnum])
2
1
3
0
4
3
5
```
So that would say the matching is:

```
rdx  2
rsi  1
r10  3
rdi  0
r8  4
rcx  3
r9  5
```
and the order is:

```
rdi
rsi
rdx
r10
r8
r9
```
I think my questions are - what about `%rcx`? I thought that should be the 4th register. And why are r8-r10 out of order? E.g., from http://6.s081.scripts.mit.edu/sp18/x86-64-architecture-guide.html if r10 is a temporary register, it looks like it's being used instead of `%rcx` and then r8 and r9, are properly used for the 5th and 6th arguments? I'm pretty new to this kind of work - thanks for answering my questions!

**Comment by rhelmot** [comment]
The thing you're missing is that once you access `cc.arch`, you're accessing an archinfo Arch object, which is information about the architecture your guest code is running on, not the actual callsite. You want `cc.arg_locs(prototype)`.

**Comment by vsoch** [comment]
ahhh so close! I tried calling that function but didn't know what to provide to it :laughing:  We have registers!

```
 for order, register in enumerate(convention.cc.arg_locs(convention.prototype)):
     ...:     print(register)
     ...:
<rdi>
<rsi>
<rdx>
<rcx>
<r8>
<r9>
[0x8]
```
So I'm looking at `cc.get_args()` and `cc.get_args_info()` and it looks like it takes the same prototype and something called a state - and I found that here! https://github.com/angr/angr/blob/master/angr/sim_state.py. Is it correct to do one of these:

```python
# is one of these correct / are they the same (what is a state?)
state = angr.SimState(arch=convention.cc.arch)
WARNING | 2022-02-04 13:25:22,905 | angr.sim_state | SimState defaulting to symbolic mode.
state = angr.SimState(arch=convention.cc.arch.name)
WARNING | 2022-02-04 13:26:00,932 | angr.sim_state | SimState defaulting to symbolic mode.
```
And then I can call both functions above:

```python
In [238]: convention.cc.get_arg_info(state=state, prototype=convention.prototype)
WARNING | 2022-02-04 13:26:28,207 | angr.storage.memory_mixins.default_filler_mixin | Filling register rdi with 8 unconstrained bytes referenced from 0x0 (unknown)
WARNING | 2022-02-04 13:26:28,209 | angr.storage.memory_mixins.default_filler_mixin | Filling register rsi with 8 unconstrained bytes referenced from 0x0 (unknown)
WARNING | 2022-02-04 13:26:28,210 | angr.storage.memory_mixins.default_filler_mixin | Filling register rdx with 8 unconstrained bytes referenced from 0x0 (unknown)
WARNING | 2022-02-04 13:26:28,211 | angr.storage.memory_mixins.default_filler_mixin | Filling register rcx with 8 unconstrained bytes referenced from 0x0 (unknown)
WARNING | 2022-02-04 13:26:28,212 | angr.storage.memory_mixins.default_filler_mixin | Filling register r8 with 8 unconstrained bytes referenced from 0x0 (unknown)
WARNING | 2022-02-04 13:26:28,213 | angr.storage.memory_mixins.default_filler_mixin | Filling register r9 with 8 unconstrained bytes referenced from 0x0 (unknown)
WARNING | 2022-02-04 13:26:28,214 | angr.storage.memory_mixins.default_filler_mixin | Filling register rsp with 8 unconstrained bytes referenced from 0x0 (unknown)
WARNING | 2022-02-04 13:26:28,681 | angr.storage.memory_mixins.default_filler_mixin | Filling memory at 0xfffffffffffffff0 with 8 unconstrained bytes referenced from 0x0 (unknown)
```
should I be worried about the warnings? I still don't see any variables names here:

```python
[(long long, 'unknown', <rdi>, <BV64 reg_rdi_8143_64{UNINITIALIZED}>),
 (long long, 'unknown', <rsi>, <BV64 reg_rsi_8144_64{UNINITIALIZED}>),
 (long long, 'unknown', <rdx>, <BV64 reg_rdx_8145_64{UNINITIALIZED}>),
 (long long, 'unknown', <rcx>, <BV64 reg_rcx_8146_64{UNINITIALIZED}>),
 (long long, 'unknown', <r8>, <BV64 reg_r8_8147_64{UNINITIALIZED}>),
 (long long, 'unknown', <r9>, <BV64 reg_r9_8148_64{UNINITIALIZED}>),
 (long long,
  'unknown',
  [0x8],
  <BV64 mem_fffffffffffffff0_8150_64{UNINITIALIZED}>)]
```
or here
```python
convention.cc.get_args(state=state, prototype=convention.prototype)
Out[239]: 
[<BV64 reg_rdi_8143_64{UNINITIALIZED}>,
 <BV64 reg_rsi_8144_64{UNINITIALIZED}>,
 <BV64 reg_rdx_8145_64{UNINITIALIZED}>,
 <BV64 reg_rcx_8146_64{UNINITIALIZED}>,
 <BV64 reg_r8_8147_64{UNINITIALIZED}>,
 <BV64 reg_r9_8148_64{UNINITIALIZED}>,
 <BV64 mem_fffffffffffffff0_8150_64{UNINITIALIZED}>]
```
will keep looking around...

**Comment by vsoch** [comment]
The model I'm working on technically (for now) includes the names of parameters because it's useful, but if they aren't available it will still work. We really need to know registers and sizes and types (which are all here!)

Is there a suggested way to recursively parse something? E.g., given a struct  or pointer, how does that work?

**Comment by rhelmot** [comment]
How do you want to get variable names? Are they in the debug info? I don't think we support extracting variable names from DWARF debugging information yet. If you have the function prototype or declaration as a string, you can parse C source into SimType objects using `angr.sim_type.parse_file()`, and that will give you a structure which contains more prototype objects, except these ones actually have names attached. You can use that to build a map from argument position to variable name.

get_args and get_args_info both are meant for FFI interaction with angr SimStates - a state is an object which represents a whole program address space, registers, and process-specific OS information (like files). I recommend reading through https://docs.angr.io/ to get a handle on angr's emulation capabilities.

I'm not super clear on what it means to parse a type besides taking its string representation and coming up with an object representing its structure, but if you're talking about navigating the SimType objects, you should get familiar with the SimType docs (http://angr.io/api-doc/angr.html#angr.sim_type.SimType) and maybe even sim_type.py - each class has attributes on it containing its children and properties.

**Comment by vsoch** [comment]
Okay trying to figure out how to parse a class (Arithmetic) I have a function with arguments:

```cpp
#include "MathLibrary.h"

namespace MathLibrary
{
    double Arithmetic::Add(double a, double b)
    {
        return a + b;
    }

    double Arithmetic::Subtract(double a, double b)
    {
        return a - b;
    }

    double Arithmetic::Multiply(double a, double b)
    {
        return a * b;
    }

    double Arithmetic::Divide(double a, double b)
    {
        return a / b;
    }
}
```
but no clear record of that - 

```python
func
<Function _ZN11MathLibrary10Arithmetic3AddEdd (0x4010fa)>

# no output below
for arg_info in convention.cc.get_arg_info(state=state, prototype=convention.prototype):
    print(arg_info)

# no locations
convention.cc.arg_locs(convention.prototype)
[]
```
I'm guessing I need to work backwards and somehow get the class type (SimType) from the function (how do I do that?) I am spending a lot of time looking at the objects and tutorials and code base, I haven't found a good example of what I want to do, sorry for asking so many questions!

**Comment by vsoch** [comment]
Ahh I'm going to try this!

https://github.com/angr/angr/blob/cb89c1978a61524fb18202f89e587f1b19a9843d/angr/analyses/class_identifier.py#L7

**Comment by vsoch** [comment]
Okay looks like there isn't a vtables list:

```
classes = self.proj.analyses.ClassIdentifier()
WARNING | 2022-02-04 15:44:06,928 | angr.analyses.vtable | VtableFinder analysis is skipped
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Input In [3], in <module>
----> 1 classes = self.proj.analyses.ClassIdentifier()

File ~/Desktop/Code/smeagle-python/env/lib/python3.8/site-packages/angr/analyses/analysis.py:116, in AnalysisFactory.__call__(self, *args, **kwargs)
    113         raise AngrAnalysisError('The "progress_callback" parameter must be a None or a callable.')
    115 oself._show_progressbar = show_progressbar
--> 116 oself.__init__(*args, **kwargs)
    117 return oself

File ~/Desktop/Code/smeagle-python/env/lib/python3.8/site-packages/angr/analyses/class_identifier.py:23, in ClassIdentifier.__init__(self)
     21 self.classes = {}
     22 vtable_analysis = self.project.analyses.VtableFinder()
---> 23 self.vtables_list = vtable_analysis.vtables_list
     24 self._analyze()

AttributeError: 'VtableFinder' object has no attribute 'vtables_list'
```
if I try to do it manually I also see there isn't one:
```
In [5]: finder = self.proj.analyses.VtableFinder()
WARNING | 2022-02-04 15:45:07,882 | angr.analyses.vtable | VtableFinder analysis is skipped

In [6]: finder.__dict__
Out[6]: 
{'named_errors': {},
 'errors': [],
 'log': [],
 '_fail_fast': False,
 '_name': 'VtableFinder',
 'project': <Project example/libmath-v1.so>,
 'kb': <angr.knowledge_base.knowledge_base.KnowledgeBase at 0x7f2798399b20>,
 '_progress_callback': None,
 '_show_progressbar': False}

In [7]: finder.vtables_list
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Input In [7], in <module>
----> 1 finder.vtables_list

AttributeError: 'VtableFinder' object has no attribute 'vtables_list'
```

**Comment by rhelmot** [comment]
The issue is reported in the warning you showed:  `angr.analyses.vtable | VtableFinder analysis is skipped`. It is because your binary doesn't have any of the sections that the VtableFinder analysis needs to do its work, that is, `[".data.rel.ro", ".rodata", ".data.rel.ro.local"]`.

I don't believe you should need to do any class analysis, I think just the normal function analyses are supposed to work on the stuff you're dealing with. It's a bug that the calling convention analysis is reporting no arguments - I suspect this is because the args are being passed through SSE registers and the analysis isn't wired to take those into account - cc @ltfish. Try rewriting your example with integers and see what happens.

**Comment by vsoch** [comment]
okay found a workaround! I created a custom analyzer that handled the missing variable

```
class CustomClassIdentifier(ClassIdentifier):
    def __init__(self):
        if "CFGFast" not in self.project.kb.cfgs:
            self.project.analyses.CFGFast(cross_references=True)
        self.classes = {}
         vtable_analysis = self.project.analyses.VtableFinder()
         vtable_analysis.analyze()
         if not hasattr(vtable_analysis, "vtables_list"):
             self.vtables_list = []
         else:
             self.vtables_list = vtable_analysis.vtables_list
         self._analyze()

AnalysesHub.register_default("CustomClassIdentifier", CustomClassIdentifier)
```

and it worked! At least I found the class!
```python
classes = self.proj.analyses.CustomClassIdentifier()
WARNING | 2022-02-04 15:53:02,441 | angr.analyses.vtable | VtableFinder analysis is skipped

classes
<CustomClassIdentifier Analysis Result at 0x7f2794169cd0>

classes.__dict__
{'named_errors': {},
 'errors': [],
 'log': [],
 '_fail_fast': False,
 '_name': 'CustomClassIdentifier',
 'project': <Project example/libmath-v1.so>,
 'kb': <angr.knowledge_base.knowledge_base.KnowledgeBase at 0x7f2798399b20>,
 '_progress_callback': None,
 '_show_progressbar': False,
 'classes': {'MathLibrary::Arithmetic': class MathLibrary::Arithmetic},
 'vtables_list': []}
```

> I don't believe you should need to do any class analysis, I think just the normal function analyses are supposed to work on the stuff you're dealing with. It's a bug that the calling convention analysis is reporting no arguments - I suspect this is because the args are being passed through SSE registers and the analysis isn't wired to take those into account - cc @ltfish. Try rewriting your example with integers and see what happens.

oh gotcha  - yes I can give that a shot!

**Comment by vsoch** [comment]
okay changed to:
```cpp
#include "MathLibraryChanged.hpp"

namespace MathLibrary
{
    int Arithmetic::Add(int a, int b)
    {
        return a + b;
    }

    double Arithmetic::Subtract(double a, double b)
    {
        return a - b;
    }

    double Arithmetic::Multiply(double a, double b)
    {
        return a * b;
    }

    double Arithmetic::Divide(double a, double b)
    {
        return a / b;
    }
}
```

but the same listings are still empty - the only thing I can get is the return type.

**Comment by vsoch** [comment]
Oh that's interesting when I parse the class via my workaround above, it looks like I do see the params!

```
In [11]: v.function_members
Out[11]: 
{4198650: MathLibrary::Arithmetic::Add(int, int)() -> None,
 4198674: MathLibrary::Arithmetic::Subtract(double, double)() -> None,
 4198704: MathLibrary::Arithmetic::Multiply(double, double)() -> None,
 4198734: MathLibrary::Arithmetic::Divide(double, double)() -> None}
```
but I don't think I can use this - it looks like that's just the label (and there aren't actually the args here).

```python
member.__dict__
{'label': 'MathLibrary::Arithmetic::Divide(double, double)',
 'args': [],
 'returnty': None,
 'arg_names': (),
 'variadic': False,
 'ctor': True,
 'dtor': False}
```

It was worth a shot! Indeed it might be a bug - I hope we can fix it or find another way?

**Comment by vsoch** [comment]
So this is kind of weird / interesting - I randomly was testing things and saw variables, and it looks like

```python
In [2]: convention.cc.get_arg_info(state=state, prototype=convention.prototype)
Out[2]: 
[(long long, 'unknown', <rdi>, <BV64 reg_rdi_33_64{UNINITIALIZED}>),
 (long long, 'unknown', <rsi>, <BV64 reg_rsi_34_64{UNINITIALIZED}>)]
```
It seems to show up sort of randomly? And I see this before it does:

```
WARNING | 2022-02-04 19:01:46,555 | angr.storage.memory_mixins.default_filler_mixin | The program is accessing memory or registers with an unspecified value. This could indicate unwanted behavior.
WARNING | 2022-02-04 19:01:46,555 | angr.storage.memory_mixins.default_filler_mixin | angr will cope with this by generating an unconstrained symbolic variable and continuing. You can resolve this by:
WARNING | 2022-02-04 19:01:46,555 | angr.storage.memory_mixins.default_filler_mixin | 1) setting a value to the initial state
WARNING | 2022-02-04 19:01:46,555 | angr.storage.memory_mixins.default_filler_mixin | 2) adding the state option ZERO_FILL_UNCONSTRAINED_{MEMORY,REGISTERS}, to make unknown regions hold null
WARNING | 2022-02-04 19:01:46,555 | angr.storage.memory_mixins.default_filler_mixin | 3) adding the state option SYMBOL_FILL_UNCONSTRAINED_{MEMORY,REGISTERS}, to suppress these messages.
WARNING | 2022-02-04 19:01:46,555 | angr.storage.memory_mixins.default_filler_mixin | Filling register rdi with 8 unconstrained bytes referenced from 0x0 (unknown)
WARNING | 2022-02-04 19:01:46,556 | angr.storage.memory_mixins.default_filler_mixin | Filling register rsi with 8 unconstrained bytes referenced from 0x0 (unknown)
```
This doesn't make sense to me, but when I added a time.sleep and ran that small loop twice:

```
            for _ in range(2):
                time.sleep(1)
                convention = self.proj.analyses.CallingConvention(func, cfg=self.cfg, analyze_callsites=True)
            
            # Run the calling convention analysis (this has arch and registers)
            # We get back a SimCC to give us detailed information about how a function's parameters should be laid out in registers and memory.
            #classes = self.proj.analyses.CustomClassIdentifier()

                # Then variable recovery
                vr = self.proj.analyses.VariableRecoveryFast(func)

                # And type inference
                types = self.proj.analyses.Typehoon(vr.type_constraints)
```
for one out of the four functions, I did see the arguments parse!

```
{
            "name": "_ZN11MathLibrary10Arithmetic3AddEii",
            "size": 24,
            "direction": "export",
            "parameters": [
                {
                    "type": "long long",
                    "size": 64,
                    "location": "rdi"
                },
                {
                    "type": "long long",
                    "size": 64,
                    "location": "rsi"
                },
                {
                    "type": "int",
                    "size": 32,
                    "location": "rax"
                }
            ]
        }
```
Does that give any hints to what is going on? Note there are both long long. If you can point me somewhere to look and maybe some thoughts I can try to help!

**Comment by vsoch** [comment]
Trying to debug the error messages a little bit - for this warning:

```
WARNING | 2022-02-04 19:34:13,203 | angr.analyses.variable_recovery.engine_base | Reading memory with overlapping variables: {(0, <0x4010b0[is_1]|s_0(stack) bp, 8 B>), (None, <0x4010b0[is_1]|s_0(stack) bp, 8 B>)}. Ignoring all but the first one.
```

At this chunk of code:

https://github.com/angr/angr/blob/cb89c1978a61524fb18202f89e587f1b19a9843d/angr/analyses/variable_recovery/engine_base.py#L564-L569

why is a variable added if the offset is None?

**Comment by ltfish** [comment]
The class identifier is very much untested. I know very little about it at this point although I will be intensively using it soon...

If you post the binary here, I can write some code (and fix some bugs) for you to achieve your purposes. For one thing, this is what you want:

```
proj.analyses.CompleteCallingConventions(cfg=cfg, recover_variables=True)
```

instead of calling `CallingConventionAnalysis` before recovering variables for each function.

> why is a variable added if the offset is None?

It's a long-standing issue (or bug) for which I did not care enough to track down the root cause. The `offset` thing in VariableRecovery also needs some proper re-thinking from my end. No promise, but if I have some cycles this weekend, I'll take a look at it.

**Comment by rhelmot** [comment]
@ltfish is the person who can answer your questions more precisely but I can tell you this: don't use get_arg_info, you never want to extract a value from a state unless you're interfacing with a running (or emulated) instance of the program, and that's part of what get_arg_info does.

**Comment by vsoch** [comment]
Awesome! Here is the binary with the class I was testing (it's just a small test library) and remove the .txt.

[libmath-v2.so.txt](https://github.com/angr/angr/files/8007302/libmath-v2.so.txt)

I am usually on my computer working on something for most of the weekend so keep me updated if there is anything I can help with! Superficially I tried what you suggested before entering the loop:

```python
conv.function_needs_variable_recovery(func)
```

and it does look like some kind of recovery is needed.

@rhelmot thank you so kindly! I'll refactor now to not use that.

**Comment by vsoch** [comment]
okay this is getting me the same result (just for 1/4 that has the int parameters) and I chose this function because it doesn't take a state.
```python
for i, register in enumerate(convention.cc.arg_locs(convention.prototype)):
        typ = convention.prototype.args[i]
        typ = typ.with_arch(convention.cc.arch)
        entry["parameters"].append(
            {
                "type": typ.c_repr(),
                "size": typ.size,
                "location": register.reg_name,
            }
        )
```

**Comment by ltfish** [comment]
The problem with angr's calling convention analysis (CCA) is that it does not yet support floating point arguments. We are in the process of adding FP support to angr decompiler, which means CCA will soon support FP arguments as well. But before that, the only thing you can do for acquiring the argument registers for functions with floating numbers as arguments is the following:

- Cheat by parsing the demangled function name of the C++ function. In your example, `project.kb.functions[0x40114e].demangled_name` returns `MathLibrary::Arithmetic::Divide(double, double)`.
- Since this function takes two `double`s as the argument, now let's see which registers they should be:

```python
from angr.sim_type import SimTypeInt, SimTypeDouble, SimTypeFunction

proto = SimTypeFunction([SimTypeDouble(), SimTypeDouble()], SimTypeInt())
regs = proj.kb.functions[0x40114e].calling_convention.arg_locs(proto)
print(regs)
```

Output:
```
[<xmm0>, <xmm1>]
```

Btw, you can parse the proto using angr:

```python
from angr.sim_type import parse_cpp_file

print(parse_cpp_file(instance.kb.functions[0x40114e].demangled_name, with_param_names=True)[0]['MathLibrary::Arithmetic::Divide'])
```

Output:
```
None(void*, double, double) -> BOT
```

**Comment by vsoch** [comment]
That did the trick! :partying_face: 

```
        {
            "name": "_ZN11MathLibrary10Arithmetic6DivideEdd",
            "size": 30,
            "direction": "export",
            "parameters": [
                {
                    "type": "void *",
                    "size": 64,
                    "location": "rdi"
                },
                {
                    "type": "double",
                    "size": 64,
                    "location": "xmm0"
                },
                {
                    "type": "double",
                    "size": 64,
                    "location": "xmm1"
                },
                {
                    "type": "int",
                    "size": 32,
                    "location": "rax"
                }
            ]
        },
```

Is there an issue I should subscribe to, to watch for updates for parsing floats?

Going to dig more into specific type parsing next (more complex examples).

**Comment by vsoch** [comment]
Found a similar issue when parsing a struct - here I have a struct with fields:

```cpp
struct Foo {
        int x;
        char c[32];
};
```

And I don't know if there is a way to just parse it? I was looking around the knowledge base and didn't see anything that looked like a struct:

```
{'_project': <Project libarray.so>,
 '_plugins': {'name': 'global',
  'cfgs': <CFGManager with 1 CFGs>,
  'functions': <angr.knowledge_plugins.functions.function_manager.FunctionManager at 0x7f99aaa71490>,
  'labels': <angr.knowledge_plugins.labels.Labels at 0x7f99aaa43580>,
  'xrefs': <angr.knowledge_plugins.xrefs.xref_manager.XRefManager at 0x7f99aaa43670>,
  'indirect_jumps': {},
  'propagations': <angr.knowledge_plugins.propagations.Propagations at 0x7f99aa8cea60>,
  'variables': <angr.knowledge_plugins.variables.variable_manager.VariableManager at 0x7f99aa975fd0>}}
```
And then I thought oh, maybe I can find it as a parameter to a function? So this:

```
void foo(Foo f){}
```

but when it parses I don't see any fields:

```
typ
struct Foo

type(typ)
angr.sim_type.SimStruct

typ.fields
OrderedDict()
```
And I was expecting to see an int and char. Based on how I'm seeing this library work - is there another analyzer that I need to run?

**Comment by vsoch** [comment]
okay I found this https://github.com/angr/angr/pull/2955/files#diff-7a0fd1f9c47e1068618ab467c30dcb6649ca71351f6d48f5141f333b8822d513R514-R520 so I don't think getting the struct as an argument (parsed) is supported yet - going to keep looking if I can parse it outside of the context of the function.

**Comment by vsoch** [comment]
@ltfish is it correct that we can't go any further (because structs and maybe others aren't parsed) or is there a workaround? Or something I can help with or debug? I made pretty good progress on Friday and the weekend (https://github.com/buildsi/smeagle-python) - and I really like the library and am hoping I can parse all the different remaining types.

**Comment by vsoch** [comment]
Just to keep you in the loop! I found this example in testing and am going to test it to see what it shows me tomorrow. https://github.com/angr/angr/blob/755dcb8b80907e95ae2d6b29c181ef9dc50193f3/tests/test_variablerecovery.py#L76-L131

**Comment by vsoch** [comment]
Ack that didn't seem to help. :/ Should I try this? https://github.com/angr/angr/issues/403#issuecomment-290302006

**Comment by vsoch** [comment]
Sure! So in simpler terms I just want to have a json object with structures and their fields, kind of like:

```json
   {
    "type": "struct",
    "name": "structHsqlmtcfjxgalrdce",
    "is_pointer": true,
    "fields": [
     {
      "name": "fpIntVztmefdjczvtfppzxgu",
      "type": "long long",
      "is_signed": true,
      "is_pointer": true,
      "value": "-2676140095681027988"
     }
    ]
   },
```
And the same for the other more complex types. Does that make sense? And apologies for my bad terminology - I definitely overusing parsing to describe too many things :)

**Comment by rhelmot** [comment]
Right, but what's your input? Is it a binary? are there debug symbols? is it source code?

**Comment by vsoch** [comment]
oh right! So we can assume that it's a binary with debug dwarf. We don't have source code.

**Comment by rhelmot** [comment]
Okay - extracting types from debug symbols is not something we have implemented in the angr suite, but it's definitely something we _want_ to have implemented - if you think the additional accuracy is worth it, I can guide you through implementing this for CLE.

**Comment by rhelmot** [comment]
Awesome! So we will be using pyelftools, that's what all of CLE's ELF support is built on.

The main challenge here will be coming up with an interface that will be appropriate to export, since we generally don't like exporting pyelftools objects directly from CLE. To be honest, I would kind of like to split angr.sim_type out into its own package that CLE can depend on, but that's a task for another day.

The place you want to start writing is [here](https://github.com/angr/cle/blob/master/cle/backends/elf/elf.py#L156). You'll want to add a method that parses all the type information into some standardized format and dumps it onto the `self` object - should be an ELF. IDEALLY, and this is a big ideally, the format it dumps out should have nothing to do with the fact that it came from an ELF, so maybe in the future if we can extract similar metadata from PE or mach-o files, we can export the same interface on them too.

**Comment by vsoch** [comment]
okay - just wanted to leave a note here - we can't get registers from the dwarf (we need to do the CallingConvention analysis) so should:

1. the cle export a basic data structure of functions / variables / and all the associated types - I'm calling this a corpus
2. There perhaps be a variant of that CallingConvention that can (given a corpus exists) populate it with registers too?

Or we can just do 1, and then rely on a custom third party client to link the two. But I'm thinking that I'll need to do something similar to get registers for non function types that currently aren't handled, so there would ideally be an analyzer to do that.

I'm going to skip registers for now and we can pick up discussion about how to do this when you have some time!

**Comment by rhelmot** [comment]
Hm? We can totally get registers from dwarf data - DW_TAG_compile_unit -> DW_TAG_subprogram -> DW_TAG_formal_paramter -> DW_AT_location has the ability to specify a register location

**Comment by vsoch** [comment]
But wouldn't that be redundant for what angr could already do and we are introducing error to derive it differently? Is that what angr does? I was under the impression there are more complicated rules for specific arches, e.g., implementing basically: https://refspecs.linuxbase.org/elf/x86_64-abi-0.99.pdf. For the C++ project I'm working on we have an allocator class https://github.com/buildsi/Smeagle/blob/main/source/parser/x86_64/allocators.hpp that keeps track of things. Is there a simpler way?

**Comment by rhelmot** [comment]
The thing is that ideally we would like to use the information exported from the debug symbols as a _replacement_ for what angr's analyses can provide, or even make angr's analyses short-circuit themselves and return the exported information, because the debug symbols provide _ground truth_.

**Comment by rhelmot** [comment]
oh, haha, if you're working with malware or otherwise obfuscated code that's definitely the case. If you've ever used IDA pro, it has this popup when you load a binary with DWARF data that gives you the option for whether to trust the debug symbols or not. The thing is, that checkbox defaults to checked ;)

**Comment by vsoch** [comment]
okay found registers! :partying_face: These are definitely "dwarfy" for now, we will definitely want to figure out some standard way to describe these across different parsers. 

```
[['DW_OP_reg5 (r5)'], ['DW_OP_reg1 (r1)'], ['DW_OP_GNU_entry_value: (DW_OP_reg5 (r5))', 'DW_OP_stack_value']]
[['DW_OP_reg4 (r4)'], ['DW_OP_reg2 (r2)'], ['DW_OP_GNU_entry_value: (DW_OP_reg4 (r4))', 'DW_OP_stack_value']]
[['DW_OP_reg1 (r1)'], ['DW_OP_reg8 (r8)'], ['DW_OP_GNU_entry_value: (DW_OP_reg1 (r1))', 'DW_OP_stack_value']]
[['DW_OP_reg2 (r2)'], ['DW_OP_reg9 (r9)'], ['DW_OP_GNU_entry_value: (DW_OP_reg2 (r2))', 'DW_OP_stack_value']]
[['DW_OP_reg8 (r8)'], ['DW_OP_breg7 (r7): 0'], ['DW_OP_GNU_entry_value: (DW_OP_reg8 (r8))', 'DW_OP_stack_value']]
['DW_OP_fbreg: 0']
[['DW_OP_addr: 2000', 'DW_OP_stack_value']]
```

**Comment by vsoch** [comment]
lol sorry I used the word parser again! I'm hopeless. I mean different binary readers.

**Comment by vsoch** [comment]
hey a question! So I'm up to parsing call sites from dwarf and it's a little hairy / unclear (e.g., I found a GNU call site tag and then parameters under that, and I parsed dwarf expressions from them and I see like:

```
['DW_OP_GNU_entry_value: (DW_OP_GNU_regval_type: 17 (xmm0) <0x2d>)']
['DW_OP_fbreg: -24', 'DW_OP_GNU_deref_type: 8 <0x2d>', 'DW_OP_GNU_convert <0x34>', 'DW_OP_GNU_convert <0x0>']
```

But i noticed you have some code that looks like it's related to call sites - https://github.com/angr/cle/blob/c8eb415ae0c5ab84dff7ab026a5222d98c3110f3/cle/backends/elf/lsda.py#L27-L35 I'm thinking this is for exception parsing but is there a way I can connect these two?

**Comment by vsoch** [comment]
okay a quick update - I stopped reading into the dwarf expression so much and just followed the abstract origin to get a function name, and parsed the location for the call site locations the same as other parameters, and I think this is reasonable for now!

```
{
            "callsite": {
                "name": "_Z3bard",
                "params": [
                    {
                        "location": "%xmm0"
                    }
                ]
            }
        },
        {
            "callsite": {
                "name": "_Z3fooi",
                "params": [
                    {
                        "location": "%rdi"
                    }
                ]
            }
        }
```
These were for the GNU_* versions of DWARF and I haven't tested other versions but will at some point. I'm interested to look into how angr is parsing lsda/exceptions so I'm going to tackle that next.
